arangod/Cluster/ActionDescription.cpp

```diff
@@ -0,0 +1,146 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2018 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Kaveh Vahedipour
+/// @author Matthew Von-Maszewski
+////////////////////////////////////////////////////////////////////////////////
+
+#include "ActionDescription.h"
+
+#include <functional>
+
+using namespace arangodb;
+using namespace arangodb::maintenance;
+
+
+/// @brief ctor
+ActionDescription::ActionDescription(
+  std::map<std::string, std::string> const& d,
+  std::shared_ptr<VPackBuilder> const p) :
+  _description(d), _properties(p) {
+  TRI_ASSERT(d.find(NAME) != d.end());
+  TRI_ASSERT(p == nullptr || p->isEmpty() || p->slice().isObject());
+}
+
+/// @brief Default dtor
+ActionDescription::~ActionDescription() {}
+
+/// @brief Does this description have a "p" parameter?
+bool ActionDescription::has(std::string const& p) const {
+  return _description.find(p) != _description.end();
+}
+
+/// @brief Does this description have a "p" parameter?
+std::string ActionDescription::operator()(std::string const& p) const {
+  return _description.at(p);
+}
+
+/// @brief Does this description have a "p" parameter?
+std::string ActionDescription::get(std::string const& p) const {
+  return _description.at(p);
+}
+
+/// @brief Get parameter
+Result ActionDescription::get(std::string const& p, std::string& r) const {
+  Result result;
+  auto const& it = _description.find(p);
+  if (it == _description.end()) {
+    result.reset(TRI_ERROR_FAILED);
+  } else {
+    r = it->second;
+  }
+  return result;
+}
+
+/// @brief Hash function
+std::size_t ActionDescription::hash() const {
+  std::string propstr;
+  for (auto const& i : _description) {
+    propstr += i.first + i.second;
+  }
+  return std::hash<std::string>{}(propstr);
+}
+
+std::size_t ActionDescription::hash(std::map<std::string, std::string> desc) {
+  std::string propstr;
+  for (auto const& i : desc) {
+    propstr += i.first + i.second;
+  }
+  return std::hash<std::string>{}(propstr);
+}
+
+/// @brief Equality operator
+bool ActionDescription::operator==(
+  ActionDescription const& other) const {
+  return _description== other._description;
+}
+
+/// @brief Get action name. Cannot throw. See constructor
```

Comment:
This code is assuming the TRI_ASSERT() in constructor has validated the input.  But the validation only occurs in MAINTAINER builds.  How about switching this from "at" to "get".  A zero length string is actually handled where a throw is not in the create() call.

Reply:
fixed